contracts/Strategy.sol

```diff
@@ -53,80 +143,224 @@ contract Strategy is BaseStrategy {
+                if (newLoose < amountToFree) {
+                    if (_profit > newLoose) {
+                        _profit = newLoose;
+                        _debtPayment = 0;
+                    } else {
+                        _debtPayment = Math.min(
+                            newLoose.sub(_profit),
+                            _debtPayment
+                        );
+                    }
+                }
+            }
+        } else {
+            //serious loss should never happen but if it does lets record it accurately
+            _loss = debt.sub(assets);
+        }
     }
 
     function adjustPosition(uint256 _debtOutstanding) internal override {
-        // TODO: Do something to invest excess `want` tokens (from the Vault) into your positions
-        // NOTE: Try to adjust positions so that `_debtOutstanding` can be freed up on *next* harvest (not immediately)
+        if (emergencyExit) {
+            return;
+        }
+
+        uint256 wantBalance = want.balanceOf(address(this));
+
+        uint256 shares = 0;
+
+        if (wantBalance > 0) {
+            (, shares) = bentoBox.deposit(
+                BIERC20(address(want)),
+                address(this),
+                address(this),
+                wantBalance,
+                0 // setting this to 0, let's the previous argument determine the deposit size
+            );
+        }
+
+        uint256 sharesInBento =
+            bentoBox.balanceOf(BIERC20(address(want)), address(this));
+
+        if (sharesInBento > 0) {
+            bentoBox.transfer(
+                BIERC20(address(want)),
+                address(this),
+                address(kashiPair),
+                sharesInBento
+            );
+
+            kashiPair.addAsset(address(this), true, sharesInBento);
+        }
     }
 
     function liquidatePosition(uint256 _amountNeeded)
         internal
         override
         returns (uint256 _liquidatedAmount, uint256 _loss)
     {
-        // TODO: Do stuff here to free up to `_amountNeeded` from all positions back into `want`
-        // NOTE: Maintain invariant `want.balanceOf(this) >= _liquidatedAmount`
-        // NOTE: Maintain invariant `_liquidatedAmount + _loss <= _amountNeeded`
-
         uint256 totalAssets = want.balanceOf(address(this));
         if (_amountNeeded > totalAssets) {
-            _liquidatedAmount = totalAssets;
-            _loss = _amountNeeded.sub(totalAssets);
+            uint256 amountToFree = _amountNeeded.sub(totalAssets);
+
+            (, uint256 lastAccrued, ) = kashiPair.accrueInfo();
+            if (block.timestamp > lastAccrued) {
+                // We need to call accrue to accurately calculate totalAssets
+                kashiPair.accrue();
+            }
+
+            uint256 deposited = estimatedTotalAssets().sub(totalAssets);
+
+            if (deposited < amountToFree) {
+                amountToFree = deposited;
+            }
+            if (deposited > 0) {
```

Comment:
isn't checking that amountToFree > 0 more intuitive?

I believe that in practical terms it is the same tho